Missing field handling in nested data

Hey,

The schema with a nested struct inside a list causes an error. Related to #257 (assuming - thanks for the quick fix!).
Here is the minimal example:

```python
    schema = pymongoarrow.api.Schema({
        '_id': bson.ObjectId,
        'test_list_struct': [{
            'field1': {
                'sub_field1': pyarrow.string(),
                'sub_field2': pyarrow.string(),
            }
        }],
    })

    temp_collection.insert_one({
        "_id": bson.objectid.ObjectId("000000000000000000000001"),
        "test_list_struct": [
            {"field1": {
                "sub_field1": "test_data",
                # "sub_field2": "test_data",  # missing and causes the error, when uncommented it works fine
            }},
            {"field1": "test_data", }
        ],
    })

    df = pymongoarrow.api.aggregate_polars_all(temp_collection._collection,
                                               schema=schema,
                                               pipeline=[],
                                               )
```

Produces:
```python
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/home/gugleta/.cache/pypoetry/virtualenvs/cxbi-LF4XS2gW-py3.12/lib/python3.12/site-packages/pymongoarrow/api.py:368: in aggregate_polars_all
    return _arrow_to_polars(aggregate_arrow_all(collection, pipeline, schema=schema, **kwargs))
/home/gugleta/.cache/pypoetry/virtualenvs/cxbi-LF4XS2gW-py3.12/lib/python3.12/site-packages/pymongoarrow/api.py:152: in aggregate_arrow_all
    return context.finish()
/home/gugleta/.cache/pypoetry/virtualenvs/cxbi-LF4XS2gW-py3.12/lib/python3.12/site-packages/pymongoarrow/context.py:50: in finish
    array_map = _parse_builder_map(self.manager.finish())
/home/gugleta/.cache/pypoetry/virtualenvs/cxbi-LF4XS2gW-py3.12/lib/python3.12/site-packages/pymongoarrow/context.py:73: in _parse_builder_map
    builder_map[key] = StructArray.from_arrays(arrs, names=names)
pyarrow/array.pxi:4108: in pyarrow.lib.StructArray.from_arrays
    ???
pyarrow/error.pxi:155: in pyarrow.lib.pyarrow_internal_check_status
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   pyarrow.lib.ArrowInvalid: Mismatching child array lengths

pyarrow/error.pxi:92: ArrowInvalid
```

Tested with version `1.6.3`
Thanks in advance!

--- a/bindings/python/pymongoarrow/lib.pyx
+++ b/bindings/python/pymongoarrow/lib.pyx
@@ -254,6 +254,12 @@ cdef class BuilderManager:
             # For list children, the nulls are stored in the parent.
             key = field.encode('utf-8')
             parent_type = self.parent_types.get(key, None)
+            # Check if the item was in our schema but never seen, and should have a parent.
+            if parent_type is None and "." in field:
+                parent_key, _, _ = field.rpartition('.')
+                self.parent_names[key] = parent_key.encode('utf-8')
+                parent_type = BSON_TYPE_DOCUMENT
+            # Add nulls according to parent type.
             if parent_type == BSON_TYPE_ARRAY:
                 continue
             if parent_type == BSON_TYPE_DOCUMENT:

--- a/bindings/python/test/test_arrow.py
+++ b/bindings/python/test/test_arrow.py
@@ -533,6 +533,49 @@ class ArrowApiTestMixin:
             expected = json.load(fid)
         assert df.to_pylist() == expected
 
+    def test_schema_arrays_of_documents_orphaned_null(self):
+        # From https://github.com/mongodb-labs/mongo-arrow/issues/265.
+        col = self.coll
+        col.delete_many({})
+        schema = Schema(
+            {
+                "_id": ObjectId,
+                "test_list_struct": [
+                    {
+                        "field1": {
+                            "sub_field1": pa.string(),
+                            "sub_field2": pa.string(),
+                        }
+                    }
+                ],
+            }
+        )
+
+        col.insert_one(
+            {
+                "_id": ObjectId("000000000000000000000001"),
+                "test_list_struct": [
+                    {
+                        "field1": {
+                            "sub_field1": "test_data",
+                        }
+                    },
+                    {
+                        "field1": "test_data",
+                    },
+                ],
+            }
+        )
+        df = aggregate_arrow_all(col, schema=schema, pipeline=[])
+        doc = df.to_pylist()[0]
+        del doc["_id"]
+        assert doc == {
+            "test_list_struct": [
+                {"field1": {"sub_field1": "test_data", "sub_field2": None}},
+                {"field1": {"sub_field1": None, "sub_field2": None}},
+            ]
+        }
+
     def test_auto_schema_nested(self):
         # Create table with random data of various types.
         _, data = self._create_nested_data()

